### ISSUE TITLE
CS8002 Referenced assembly does not have a strong name triggers when it doesn't need to.

### ISSUE THREAD

**Issue by blowdart** [dotnet/roslyn#76197]
**Steps to Reproduce**:

1. Create a .NET 8.0 project
2. Turn on all analyzers
3. Add a nuget package or reference to an assembly that is not strong named.
4. Watch CS8002 light up.

A minimal repro, with source-code provided, is ideal.  Most compiler/language issues can be distilled into a snippet that can be pasted into [sharplab](https://sharplab.io/).

**Diagnostic Id**:

`CS8002: Referenced assembly 'whatever, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' does not have a strong name.`

**Expected Behavior**:

According to [the strong name docs](https://learn.microsoft.com/en-us/dotnet/standard/library-guidance/strong-naming)

> _Strong naming has no benefits on .NET Core/5+. C# compiler produces CS8002 warning for strong-named assemblies referencing non-strong named assemblies. It is fine to suppress this warning for libraries that target .NET Core/5+ only._

I would expect if I am targeting a runtime later than v5 (it's v8 in my project) the analyzer doesn't trigger at all.

A less optimal fix would be when creating a project targeting >= .net 5 you automatically add the lines to disable it in the project.

(For those searching for what to do about CS8002 add the following to your project if you're targeting a runtime of >= .net 5)

```xml
  <PropertyGroup>
    <!-- Disable strongname warning as that's not been enforced since .net 5.-->
    <NoWarn>$(NoWarn);CS8002</NoWarn>
  </PropertyGroup>
```

**Actual Behavior**:

The analyzer triggers, and I must supress it, because in release builds I have all warnings configured as errors.

**Comment by blowdart** [comment]
Congrats @RikkiGibson 😈

**Comment by RikkiGibson** [comment]
Is there a need to fix this in unsupported .NET versions (Core 3.1, 5)?

**Comment by jaredpar** [comment]
I think that it's fine to just do this on `net5.0` and higher. As you mentioned, earlier runtimes are well out of support. Customers can always disable the warning if they want in those environments.

**Comment by RikkiGibson** [comment]
I created the following project to try and repro the issue: https://github.com/RikkiGibson/repro-roslyn-76197

Here `consumer/` is referencing a project which is not strong named. This could be manually changed to a raw `Reference` for purposes of experimentation. But it seems like a proper `ProjectReference` should work just as well. Both are "just references" to `csc` anyway.

```xml
  <ItemGroup>
    <Reference Include="reference">
      <HintPath>../reference/bin/Debug/net8.0/reference.dll</HintPath>
    </Reference>
  </ItemGroup>
```

However I wasn't able to repro the issue. It seems like for the warning in the issue description to occur, the project being built must have strong naming enabled. Is there some step I missed which ends up causing that to get switched on? I wasn't sure how to "turn on all analyzers", possibly that is related.

**Comment by jaredpar** [comment]
> Source-build builds the package from source and it does not get strong-name signed, by design.

That seems like the wrong design then. The binaries that come out of source build should be functionally equivalent to those from official builds. I know it's an inconvenient fact but strong naming impacts scenarios like this.

**Comment by NikolaMilosavljevic** [comment]
@mmitche

**Comment by mmitche** [comment]
> > Source-build builds the package from source and it does not get strong-name signed, by design.
> 
> That seems like the wrong design then. The binaries that come out of source build should be functionally equivalent to those from official builds. I know it's inconvenient but strong naming impacts scenarios like this.
> 
> At the very least we need to be consistent across the stack in our choices here.

This isn't an option. The issue is that strong name signing requires use of weak crypto which is disabled on some platforms. RedHat did some work to disable the need to SN sign when building in their environment:

https://github.com/dotnet/runtime/issues/65874 for the runtime discussion about this, with https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Arcade.Sdk/tools/StrongName.targets#L7-L13

So really, if we wanted to be consistent between our distro partner source-only builds and MS builds, the only way to go is towards no strong naming.

**Comment by baronfel** [comment]
I'd just as soon have the SDK append to nowarns if possible, since it has the knowledge of TFM. Writing an issue for this now.

**Comment by jkotas** [comment]
> That seems like the wrong design then. The binaries that come out of source build should be functionally equivalent to those from official builds. I know it's inconvenient but strong naming impacts scenarios like this.

There is difference between having a strong name and being actually strong name signed. Binaries that come out of source build do have strong name, but they are not actually signed (they use "public signing"). This difference is not user observable via .NET APIs.  It is only observable if somebody tries to actually look at the binaries and verify their strong name signatures, but there are number of differences at this level between Microsoft build and distro build.

**Comment by jaredpar** [comment]
> I'd just as soon have the SDK append to nowarns if possible, since it has the knowledge of TFM.

I worry about that being a bit fragile cause anyone can reset `<NoWarn>` without using `$(NoWarn);` in the value and suddenly run into a confusing warning. Yes it will work broadly but could still end up in corner cases.

It's true that the compiler doesn't know about TFM but we could look to corelib for a _marker_ that identifies it as not caring about strong naming. For example: could say that DIM flag means we're on a platform that doesn't care about strong naming hence disable CS8002. That wouldn't fix anything before `net6.0` but going forward it would solve the issue.

**Comment by baronfel** [comment]
Fair points, I am biasing a bit for the 80/20 solution. I think to do it for real we'd want to stop using `NoWarn` directly in the SDK logic/targets and instead have a new, 'internal' version of the thing that we combine + distinct the values of with the user-visible `NoWarn`. We've done this strategy a few times in the past but it can be high-touch, especially because so many other tools try to call the `Csc` Task.

**Comment by jaredpar** [comment]
If we did the `<NoWarn>` approach I think that `Microsoft.CSharp.Core.targets` is the appropriate place. There are already platform specific `<NoWarn>` entries there, this would just be another one. 

The downsides of this would be: 

1. Tools directly calling `<Csc>` task. This is an anti-pattern that already requires the author to do _upkeep_ work on. Low concern about this. 
2. Customers who desire to have the `CS8002` warning on .NET Core. Hopefully this is just a hypothetical concern.

**Comment by jjonescz** [comment]
@jaredpar should this issue be closed as completed?

**Comment by jaredpar** [comment]
Yes. Fixed via #80596
